### PR TITLE
Avoid cloning elements in AppBar

### PR DESCRIPTION
**Pull request by djhi**
- [x] Avoid cloning elements in AppBar
- [x] Automatically translate MenuItemLink primaryText
- [x] Documentation
- [x] Remove the `logout` prop

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -1413,6 +1413,49 @@ const PostShow = () => (
...
+            to="/configuration"
+            primaryText="pos.configuration"
+            leftIcon={<SettingsIcon />}
+            sidebarIsOpen
+        />
+    </UserMenu>
+)
+```
+
+## `<MenuItemLink>` Automatically Translate `primaryText`
```
Comment (WiXSL) [inline UPGRADE.md:1438]:
```suggestion
## `<MenuItemLink>` Automatically Translates `primaryText`
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: UPGRADE.md [hunk 2]
```diff
@@ -1413,6 +1413,49 @@ const PostShow = () => (
...
+When a React element was provided as the `userMenu` prop, the `<AppBar>` used to clone it and inject the `logout` prop (a React element). This is no longer the case and if you provided a custom user menu, you now have to include the logout yourself.
+
+```diff
+-import { UserMenu } from 'react-admin';
++import { Logout, UserMenu } from 'react-admin';
+
+-export const MyUserMenu = (props) => (
++export const MyUserMenu = () => (
+-    <UserMenu {...props}>
++    <UserMenu {...props} logout={<Logout />}>
```
Comment (WiXSL) [inline UPGRADE.md:1427]:
```suggestion
+    <UserMenu logout={<Logout />}>
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: UPGRADE.md (hunk 2 above)
Comment (WiXSL) [reply UPGRADE.md:1427]: Or maybe is `+export const MyUserMenu = (props) => (`

File: UPGRADE.md [hunk 3]
```diff
@@ -1413,6 +1413,49 @@ const PostShow = () => (
...
+## No More Props Injection In `<AppBar>`
+
+When a React element was provided as the `userMenu` prop, the `<AppBar>` used to clone it and inject the `logout` prop (a React element). This is no longer the case and if you provided a custom user menu, you now have to include the logout yourself.
+
+```diff
+-import { UserMenu } from 'react-admin';
++import { Logout, UserMenu } from 'react-admin';
+
+-export const MyUserMenu = (props) => (
++export const MyUserMenu = () => (
```
Comment (WiXSL) [inline UPGRADE.md:1425]:
```suggestion
```

File: UPGRADE.md [hunk 4]
```diff
@@ -1413,6 +1413,49 @@ const PostShow = () => (
...
 
+## No More Props Injection In `<AppBar>`
+
+When a React element was provided as the `userMenu` prop, the `<AppBar>` used to clone it and inject the `logout` prop (a React element). This is no longer the case and if you provided a custom user menu, you now have to include the logout yourself.
+
+```diff
+-import { UserMenu } from 'react-admin';
++import { Logout, UserMenu } from 'react-admin';
+
+-export const MyUserMenu = (props) => (
```
Comment (WiXSL) [inline UPGRADE.md:1424]:
```suggestion
export const MyUserMenu = (props) => (
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — APPROVED** [review]

File: UPGRADE.md [hunk 5]
```diff
@@ -1413,6 +1413,48 @@ const PostShow = () => (
...
 
+## No More Props Injection In `<AppBar>`
+
+When a React element was provided as the `userMenu` prop, the `<AppBar>` used to clone it and inject the `logout` prop (a React element). This is no longer the case and if you provided a custom user menu, you now have to include the logout yourself.
+
+```diff
+-import { UserMenu } from 'react-admin';
++import { Logout, UserMenu } from 'react-admin';
+
+export const MyUserMenu = (props) => (
```
Comment (djhi) [inline UPGRADE.md:1424]: Actually no, we don't need the props at all now

File: UPGRADE.md [hunk 6]
```diff
@@ -1421,8 +1421,7 @@ When a React element was provided as the `userMenu` prop, the `<AppBar>` used to
 -import { UserMenu } from 'react-admin';
 +import { Logout, UserMenu } from 'react-admin';
 
--export const MyUserMenu = (props) => (
-+export const MyUserMenu = () => (
+export const MyUserMenu = (props) => (
 -    <UserMenu {...props}>
 +    <UserMenu {...props} logout={<Logout />}>
```
Comment (WiXSL) [inline UPGRADE.md:1426]:
```suggestion
+    <UserMenu logout={<Logout />}>
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — APPROVED** [review]

File: UPGRADE.md [hunk 7]
```diff
@@ -1413,6 +1413,49 @@ const PostShow = () => (
...
+When a React element was provided as the `userMenu` prop, the `<AppBar>` used to clone it and inject the `logout` prop (a React element). This is no longer the case and if you provided a custom user menu, you now have to include the logout yourself.
+
+```diff
+-import { UserMenu } from 'react-admin';
++import { Logout, UserMenu } from 'react-admin';
+
+-export const MyUserMenu = (props) => (
++export const MyUserMenu = () => (
+-    <UserMenu {...props}>
++    <UserMenu logout={<Logout />}>
```
Comment (fzaninotto) [inline UPGRADE.md:1427]: Isn't it better to invite users to add the logout element as a child of the UserMenu?

File: UPGRADE.md [hunk 8]
```diff
@@ -1413,6 +1413,49 @@ const PostShow = () => (
...
+You can pass a translation key directly as the `primaryText` prop for `<MenuItemLink>`.
+
+```diff
+const CustomUserMenu = (props: any) => {
+-    const translate = useTranslate();
+    return (
+        <UserMenu {...props}>
+            <MenuItemLink
+                to="/configuration"
+-                primaryText={translate('pos.configuration')}
```
Comment (fzaninotto) [inline UPGRADE.md:1449]: what's the rationale behind this change? It doesn't save any cloneElement.

File: packages/ra-ui-materialui/src/layout/MenuItemLink.tsx [hunk 9]
```diff
@@ -113,20 +110,28 @@ export const MenuItemLink = forwardRef((props: MenuItemLinkProps, ref) => {
             >
                 {leftIcon && (
                     <ListItemIcon className={MenuItemLinkClasses.icon}>
-                        {cloneElement(leftIcon, {
-                            titleAccess: primaryText,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/MenuItemLink.tsx:117]: this is bad for accessibility, isn't it?

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-ui-materialui/src/layout/MenuItemLink.tsx (hunk 9 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/layout/MenuItemLink.tsx:117]: Right, we should add a visually hidden text next to it

File: UPGRADE.md (hunk 8 above)
Comment (djhi) [reply UPGRADE.md:1449]: Nothing, we just avoid having to get `useTranslate` and call `translate` ourselves

File: UPGRADE.md (hunk 7 above)
Comment (djhi) [reply UPGRADE.md:1427]: It would probably be better. Requires a bit more refactoring to get rid of the logout prop alltogether

File: packages/ra-ui-materialui/src/layout/MenuItemLink.tsx (hunk 9 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/layout/MenuItemLink.tsx:117]: Acutally it's not as we already have the text just after. It's only visually hidden when the sidebar is closed but still accessible by assistive technologies

File: examples/simple/src/Layout.tsx [hunk 10]
```diff
@@ -1,21 +1,29 @@
...
+        const { onClose } = useUserMenu();
 
         return (
             <MenuItem
                 ref={ref}
                 sx={{ color: 'text.secondary' }}
                 onClick={event => {
                     setLocale(locale === 'en' ? 'fr' : 'en');
-                    props.onClick(event);
+                    onClose();
```
Comment (fzaninotto) [inline examples/simple/src/Layout.tsx:26]: we need to document this new hook and update the doc for custom menus

File: packages/ra-ui-materialui/src/layout/UserMenu.tsx [hunk 11]
```diff
@@ -23,14 +57,13 @@ export const UserMenu = (props: UserMenuProps) => {
...
     } = props;
 
-    if (!logout && !children) return null;
-    const open = Boolean(anchorEl);
-
     const handleMenu = event => setAnchorEl(event.currentTarget);
-    const handleClose = () => setAnchorEl(null);
+    const handleClose = useCallback(() => setAnchorEl(null), []);
+    const context = useMemo(() => ({ onClose: handleClose }), [handleClose]);
+    if (!children) return null;
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/layout/UserMenu.tsx:65]: By default, it should show the logout button.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: examples/simple/src/Layout.tsx (hunk 10 above)
Comment (djhi) [reply examples/simple/src/Layout.tsx:26]: Yes, that's why I tagged the PR as `WIP` and added a task item in the PR description todo ;-)

File: packages/ra-ui-materialui/src/layout/UserMenu.tsx (hunk 11 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/layout/UserMenu.tsx:65]: Unless a userMenu is provided, the `Appbar` pass its `logout` prop as the UserMenu only child (see https://github.com/marmelab/react-admin/pull/7214/files#diff-a81ad0901aeb0ca1ae0d7907215277afd74ce5e7c9691d1c7c67d1ca7d9952ccR100). Isn't it enoug?

Comment (djhi) [reply packages/ra-ui-materialui/src/layout/UserMenu.tsx:65]: Maybe we could remove this `logout` prop alltogether btw

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/UserMenu.tsx:65]: If AppBar renders UserMenu with LogoutButton as child, it overrides the actual children of the user provided UserMenu. So I'm afraid we must only do that uf UserMenu doesn't have any child yet.

Comment (djhi) [reply packages/ra-ui-materialui/src/layout/UserMenu.tsx:65]: > **Unless a userMenu is provided**

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/layout/UserMenu.tsx:65]: You're right, I'm tired. Sorry for the noise.

File: docs/Admin.md [hunk 12]
```diff
@@ -38,7 +38,6 @@ Here are all the props accepted by the component:
 - [`theme`](#theme)
 - [`layout`](#layout)
 - [`loginPage`](#loginpage)
-- [`logoutButton`](#logoutbutton)
```
Comment (WiXSL) [inline docs/Admin.md:41]: The removal of the `logoutButton` is not mentioned in the Upgrade guide

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: docs/Theming.md [hunk 13]
```diff
@@ -984,16 +1022,22 @@ const App = () => (
 
 ### Changing the Icon
 
-It is possible to use a completely [custom logout button](./Admin.md#logoutbutton) or you can simply override some properties of the default button. If you want to change the icon, you can use the default `<Logout>` component and pass a different icon as the `icon` prop.
+It is possible to use a completely [custom logout button](./Theming.md#customizing-the-logout-component) or you can simply override some properties of the default button. If you want to change the icon, you can use the default `<Logout>` component and pass a different icon as the `icon` prop.
```
Comment (WiXSL) [inline docs/Theming.md:1025]: This link doesn't exist [custom logout button](./Theming.md#customizing-the-logout-component)
I'm not sure which is the correct one

**Review by WiXSL — CHANGES_REQUESTED** [review]

File: UPGRADE.md [hunk 14]
```diff
@@ -1454,6 +1454,104 @@ export const PostEdit = () => (
...
+import ListItemIcon from '@mui/material/ListItemIcon';
+import ListItemText from '@mui/material/ListItemText';
+import SettingsIcon from '@mui/icons-material/Settings';
+
+-const ConfigurationMenu = (props) => {
+// It's important to pass the ref to allow MaterialUI to manage the keyboard navigation
++const ConfigurationMenu = forwardRef((props, ref) => {
++    const { onClose } = useUserMenu();
+    return (
+        <MenuItem
```
Comment (fzaninotto) [inline UPGRADE.md:1507]: there should be a -MenuItemLink

File: UPGRADE.md [hunk 15]
```diff
@@ -1454,6 +1454,104 @@ export const PostEdit = () => (
...
+
+When a React element was provided as the `userMenu` prop, the `<AppBar>` used to clone it and inject the `logout` prop (a React element). This is no longer the case and if you provided a custom user menu, you now have to include the logout yourself.
+
+Besides, the `<UserMenu>` used to clone its children to inject the `onClick` prop, allowing them to close the menu. It now provides a `onClose` function through a new `UserContext` accessible by calling the `useUserMenu` hook.
+
+Finally, the `<UserMenu>` no longer accepts a `logout` prop. Instead, you should pass the `<Logout>` component as one of the `<UserMenu>` children.
+
+```diff
+-import { MenuItemLink, UserMenu } from 'react-admin';
++import { Logout, MenuItemLink, UserMenu, useUserMenu } from 'react-admin';
```
Comment (fzaninotto) [inline UPGRADE.md:1495]: unused MenuItemLink

File: docs/Admin.md [hunk 16]
```diff
@@ -349,24 +345,9 @@ const App = () => (
 
 You can also disable it completely along with the `/login` route by passing `false` to this prop.
 
-See The [Authentication documentation](./Authentication.md#customizing-the-login-and-logout-components) for more details.
```
Comment (fzaninotto) [inline docs/Admin.md:352]: why did you remove this link?.

File: docs/Authentication.md [hunk 17]
```diff
@@ -936,10 +950,9 @@ import * as React from "react";
 import { Admin } from 'react-admin';
```
Comment (fzaninotto) [inline docs/Authentication.md:950]: You need to change the introductory sentence 4 lines above about LogoutButton being passed to `Admin`

File: docs/Theming.md [hunk 18]
```diff
@@ -395,26 +395,66 @@ export default MyLayout;
...
-        to="/configuration"
-        primaryText="Configuration"
-        leftIcon={<SettingsIcon />}
-        onClick={onClick} // close the menu on click
-    />
-));
+// It's important to pass the ref to allow MaterialUI to manage the keyboard navigation
+const ConfigurationMenu = React.forwardRef((props, ref) => {
+    return (
+        <MenuItemLink
```
Comment (fzaninotto) [inline docs/Theming.md:410]: We should not encourage using MenuItemLink here

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: UPGRADE.md (hunk 14 above)
Comment (djhi) [reply UPGRADE.md:1507]: Nope, that's a MenuItem for use inside a `UserMenu` and we decided that `MenuItemLink` should only be used inside `Menu` because they toggle the Sidebar on mobile

File: docs/Admin.md (hunk 16 above)
Comment (djhi) [reply docs/Admin.md:352]: I didn't, I just separated the linked paragraph in two distinct ones

File: docs/Theming.md [hunk 19]
```diff
@@ -984,16 +1022,22 @@ const App = () => (
 
 ### Changing the Icon
```
[the remark this answers is not in the document]
Comment (djhi) [reply docs/Theming.md:1023]: Should I remove this section alltogether ?

File: docs/Authentication.md (hunk 17 above)
Comment (djhi) [reply docs/Authentication.md:950]: It's done

File: UPGRADE.md (hunk 14 above)
Comment (djhi) [reply UPGRADE.md:1507]: Sorry I misread. I still don't think so. This is how it should have been written in the first place

**Review by WiXSL — APPROVED** [review]

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: UPGRADE.md (hunk 14 above)
Comment (djhi) [reply UPGRADE.md:1507]: The link you send is the `Menu` documentation, here we're dealing with the `UserMenu`.

However, there is also a section in the documentation for the `UserMenu` that use the `MenuItemLink` as well. I'll add a section for this too

Comment (djhi) [reply UPGRADE.md:1508]: Done
